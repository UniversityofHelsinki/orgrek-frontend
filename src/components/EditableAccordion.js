import React, { useState } from 'react';
import { Accordion, AccordionDetails, AccordionSummary } from './Accordion';
import Typography from '@mui/material/Typography';
import { useTranslation } from 'react-i18next';
import Placeholder from './Placeholder';

const EditableAccordion = ({
  children,
  title,
  empty = false,
  placeholder,
  modified = false,
  onChange = () => {},
  ...props
}) => {
  const { t } = useTranslation();

  // TODO: initialState of expanded should be !empty
  // However currently it does work work correctly because when changing
  // the current node, this component gets rendered first time with data of the
  // previous node, so the initialState would reflect the emptiness of that
  // node. To fix this, there should be a way to check if node details are being
  // currently fetched, and only render this component after the current data
  // is available. Migrating to Redux Toolkit Query would probably make it
  // much easier.
  const [expanded, setExpanded] = useState(true);

  const handleChange = (event, newValue) => {
    if (modified) {
      return;
    }

    setExpanded(newValue);
    onChange(event, newValue);
  };

  const modifiedText = modified && (
    <Typography variant="body1" component="span" color="text.secondary" ml={3}>
      {t('accordion.modifiedSaveBeforeClosing')}
    </Typography>
  );

  const summary = (
    <AccordionSummary
      disabled={modified}
      sx={{
        ...(modified && {
          '&.Mui-disabled': {
            opacity: 1,
          },
        }),
      }}
    >
      {title}
      {modifiedText}
    </AccordionSummary>
  );

  return (
    <Accordion
      {...props}
      expanded={modified || expanded}
      onChange={handleChange}
    >
      {summary}
      <AccordionDetails>
        <Placeholder empty={empty} placeholder={placeholder}>
          {children}
        </Placeholder>
      </AccordionDetails>
    </Accordion>
  );
};

export default EditableAccordion;
