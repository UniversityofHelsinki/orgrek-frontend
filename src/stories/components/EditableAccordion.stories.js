import React from 'react';
import EditableAccordion from '../../components/EditableAccordion';

export default {
  component: EditableAccordion,
  tags: ['autodocs'],
  argTypes: {
    onChange: { action: true, control: 'none' },
  },
  parameters: {
    docs: {
      description: {
        component:
          'Accordion with a modified indicator. Prevents collapsing when modified.',
      },
    },
  },
};

export const Default = {
  args: {
    title: 'Title',
    defaultModified: false,
    disabled: false,
  },
  render: (args) => (
    <EditableAccordion {...args}>Editable content</EditableAccordion>
  ),
};

export const Modified = {
  ...Default,
  args: {
    ...Default.args,
    defaultModified: true,
  },
};
