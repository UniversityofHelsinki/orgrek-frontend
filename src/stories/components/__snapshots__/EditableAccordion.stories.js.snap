// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/EditableAccordion Collapsed play-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "focused": true,
      "name": "Title",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
  ],
  "name": "stories-components-editableaccordion--collapsed",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/EditableAccordion Default play-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "expanded": true,
      "name": "Title",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Editable content",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-editableaccordion--default",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/EditableAccordion Modified play-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "disabled": true,
      "expanded": true,
      "name": "TitleMuokattu, tallenna muutokset ennen sulkemista",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Editable content",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-editableaccordion--modified",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;
